### ISSUE TITLE
VersionedIntervalTimeline performance corner case with high number of overlapping segments

### ISSUE THREAD

**Issue by himanshug** [apache/incubator-druid#8846]
Recently all of our historical node restarted on a cluster which was serving about 50000 segments of following nature.
each segment's interval would be 24 hours and each successive segment overlaps with previous one for 1439 minutes (1440 minutes is 24 hours), for example segment intervals and versions might look like

2019-01-01T00:00:00.000Z - 2019-01-01T23:59:00.000Z , v1
2019-01-01T00:01:00.000Z - 2019-01-02T00:00:00.000Z , v2
2019-01-01T00:02:00.000Z - 2019-01-02T00:01:00.000Z , v3
2019-01-01T00:03:00.000Z - 2019-01-02T00:02:00.000Z , v4
...
...

that triggered a sequence of `VersionIntervalTimeline.remove(..)` calls for each segment one by one and broker/coordinator never recovered and needed a forced restart because `VersionIntervalTimeline.remove(..)` becomes very expensive for above scenario and never finished.

I did a quick prototype to batch multiple `VersionIntervalTimeline.remove(..)` calls into a single `VersionIntervalTimeline.removeAll(..)` call which could be used when data servers go down which had few optimizations possible. Batched call would first remove all entries `allTimelineEntries` and then from `complete/incompletePartitionTimeline` and then adjust them based on the state of  `allTimelineEntries` ,  with batched version `allTimelineEntries` has significantly fewer entries and no unnecessary corrections are to be made to `complete/incompletePartitionTimeline` which happens in non-batched removals.

..creating this issue to discuss other proposed solutions.

**Comment by jihoonson** [comment]
I like this idea. @himanshug are you considering making a PR?

**Comment by himanshug** [comment]
@jihoonson thanks for triggering the discussion :)

I faced this issue because of the very specific way  segment intervals overlapped and that all historicals  were restarted simultaneously. This issue disappears if historicals are restarted one by one. That is what I opted as a work around to solve this problem because bringing down all historicals in prod simultaneously wasn't going to give customers a good experience anyway. Hence, this issue lost its importance somewhat. Did you face this in another context ?

also, there is a caveat (on broker) and hence the approach needed discussion, operations in `VersionedIntervalTimeline` need to hold a  lock to  protect all data structures  .. currently lock is not held for long as we do things  one segment at a time .. this gives incoming queries a  chance to lookup segments and execute . With batching,  lock  could potentially be held for relatively longer and queries might block.

Since the issue doesn't appear in regular druid deployment/upgrade use case but in situations where users is doing something major on druid cluster requiring bringing down all historicals, they could always just restart the broker/coordinator in the end as well. For this reason, I kept my prototype in prototype stage and didn't create a PR.

That said, if this issue is surfacing in other use cases giving pain then it might be worth doing the batching. I could probably dig out the code and do a PR. However, at this point, I don't think it is necessary. 

PS: I kept this issue open just in case someone else noticed this issue in another context.

**Comment by jihoonson** [comment]
Ah, I'm looking at the performance of `VersionedIntervalTimeline` using `VersionedIntervalTimelineBenchmark` and the issue I'm seeing is the `remove()` method is too slow. From what I've seen, it is slow because it needs to determine the new visible segments after a segment is removed and I guess `removeAll()` could be faster if the new visible segments can be determined at once after removing segments is done.

> also, there is a caveat (on broker) and hence the approach needed discussion, operations in VersionedIntervalTimeline need to hold a lock to protect all data structures .. currently lock is not held for long as we do things one segment at a time .. this gives incoming queries a chance to lookup segments and execute . With batching, lock could potentially be held for relatively longer and queries might block.

This is true, so I think we can keep both `remove()` and `removeAll()` so that callers can use either of them based on their preferences. AFAIT, concurrent access to the timeline is required only in brokers, so maybe the broker can use `remove()` to reduce the lock contention while others use `removeAll()`. Does it make sense?

**Comment by himanshug** [comment]
Yeah, coordinator can probably use batching  without any downsides considering coordinator helpers run serially. Please  feel free to take it up if you want.
